dpkg_lowpkg: Do not access dpkg internals directly

bdrung opened the pull request:
salt contains modules, which directly access the dpkg internal database, instead of using one of the public interfaces provided by dpkg. This is a problem for several reasons, because even though the layout and format of the dpkg database is administrator friendly, and it is expected that those might need to mess with it, in case of emergency, this “interface” does not extend to other programs besides the dpkg suite of tools. The admindir can also be configured differently at dpkg build or run-time. And finally, the contents and its format, will be changing in the near future.

So use something like:

```
dpkg-query --showformat '${db-fsys:Last-Modified}\n' --show $pkg
```

to get the mtime from /var/lib/dpkg/info/<package>.list files.

`/var/lib/dpkg/available` should not be read directly. Instead `dpkg-query --print-avail` should be used. The `/var/lib/dpkg/available` file is only kept up-to-date when using dselect, but nowadays `apt` is used, and therefore this file does not provide much more information than `dpkg-query -W`. Users of APT-based frontends should use `apt show` or in our case the `pkg.show` salt module.

This merge request addresses the issues in `dpkg_lowpkg`. A separate merge request for `alternatives` will follow. Please have a look at the individual commits for more details of the changes.

Bug: https://github.com/saltstack/salt/issues/52605 and drive by fix for #58735
Bug-Debian: https://bugs.debian.org/944970

cmcmarrow commented:
@bdrung thanks for the fix!

bdrung commented:
Rebased on master.

bdrung commented:
Any updates on this? maybe @Ch3LL?

Ch3LL commented:
Looks like some of the test failures are related. I can give a review once they are cleaned up. Thanks

twangboy reviewed (changes requested):
This needs a changelog. The test needs to be migrated to pytests. This needs to be rebased and conflicts resolved.

bdrung committed:
dpkg_lowpkg: Do not access dpkg internals directly

Fixes #52605.

`salt.modules.dpkg_lowpkg` previously read `/var/lib/dpkg/available`
and `/var/lib/dpkg/info/<package>.list` directly. The dpkg database
layout is an internal interface; lintian flags this via the
`uses-dpkg-database-directly` tag. Switch to dpkg-query exclusively:

* `info()` derives the package install date from the dpkg field
  `${db-fsys:Last-Modified}` (available since dpkg 1.19.3, shipped in
  Debian 10 / Ubuntu 18.04) instead of reading the mtime of
  `/var/lib/dpkg/info/<package>.list`. Older dpkg returns an empty
  value and Salt simply omits `install_date` from the result rather
  than crashing.
* `_get_pkg_ds_avail()` now invokes `dpkg-query --print-avail`
  (when dselect is installed) instead of reading
  `/var/lib/dpkg/available`. The merged-extra-keys behaviour for
  dselect-using systems is preserved; the on-disk source is the
  only thing that changes.
* `_get_pkg_info()` passes its dpkg-query invocation to `cmd.run_all`
  as a list (no shell), uses Python 3.12+-safe
  `datetime.fromtimestamp(..., tz=timezone.utc)` for the install_date
  conversion, and no longer requests the always-empty
  `${binary:Revision}`, `${Summary}`, `${SHA1}`, `${SHA256}` fields
  (these substitution variables either do not exist for the package
  database or are not populated, so the values were already always
  empty and stripped before reaching consumers).

The `installed_size` key continues to be requested from dpkg-query and
then filtered out by `info()`'s existing technical-keys block, so the
public return shape of `lowpkg.info()` is unchanged for the common
case. Consumers using dselect retain the merged extra fields.

Adds regression tests pinning: the command is passed as a list (not a
shell string); the format string requests the public
`${db-fsys:Last-Modified}` field; no command argument references
`/var/lib/dpkg/info` or `/var/lib/dpkg/available`; older-dpkg empty
install_date is tolerated; `_get_pkg_ds_avail()` calls
`dpkg-query --print-avail` only when dselect is installed.

Signed-off-by: Benjamin Drung <benjamin.drung@ionos.com>
